internal/tempfile/tmpfile_test.go

```diff
@@ -0,0 +1,173 @@
+package tempfile_test
+
+import (
+	"os"
+	"path/filepath"
+	"runtime"
+	"strings"
+	"testing"
+
+	"github.com/buildkite/agent/v3/internal/tempfile"
+	"gotest.tools/v3/assert"
+)
+
+func TestNew(t *testing.T) {
+	t.Parallel()
+
+	for _, test := range []struct {
+		name  string
+		opts  []tempfile.TempFileOpts
+		check func(t *testing.T, f *os.File)
```

Comment:
This is an interesting approach to the problem of different test cases requiring different logic to check. I'm not super opposed to it, because it has the interesting property that test-case-specific checking logic is _right there_ in the test case. But then... would it read better if done as separate tests or non-table-driven subtests?

https://google.github.io/styleguide/go/decisions#table-driven-tests says:

> When some test cases need to be checked using different logic from other test cases, it is more appropriate to write multiple test functions, as explained in [GoTip #50: Disjoint Table Tests](https://google.github.io/styleguide/go/index.html#gotip). The logic of your test code can get difficult to understand when each entry in a table has its own different conditional logic to check each output for its inputs. If test cases have different logic but identical setup, a sequence of [subtests](https://google.github.io/styleguide/go/decisions#subtests) within a single test function might make sense.

I will also point out that, if each case was separated out (i.e. not table driven), there'd be no need for `opts`, which appears to be the only use of `TempFileOpts` outside of `tempfile`, so you could avoid exporting it.